tableaudocumentapi/datasource.py

```diff
@@ -258,3 +261,84 @@ def _get_metadata_objects(self):
         return [_column_object_from_column_xml(self._datasourceTree, xml)
                 for xml in self._datasourceTree.findall('.//column')]
+
+    def add_field(self, name, datatype, role, type, caption):
+        """ Adds a base field object with the given values.
+
+        Args:
+            name: Name of the new Field. String.
+            datatype:  Datatype of the new field. String.
+            role:  Role of the new field. String.
+            type:  Type of the new field. String.
+            caption:  Caption of the new field. String.
+
+        Returns:
+            The new field that was created. Field.
+        """
+        # TODO: A better approach would be to create an empty column and then
+        # use the input validation from its "Field"-object-representation to set values.
+        # However, creating an empty column causes errors :(
+
+        # If no caption is specified, create one with the same format Tableau does
+        if not caption:
+            caption = name.replace('[', '').replace(']', '').title()
+
+        # Create the elements
+        column = ET.Element('column')
+        column.set('caption', caption)
+        column.set('datatype', datatype)
+        column.set('role', role)
+        column.set('type', type)
+        column.set('name', name)
+
+        self._datasourceTree.getroot().append(column)
+
+        # Refresh fields to reflect changes and return the Field object
+        self._refresh_fields()
+        return self.fields[name]
+
+    def remove_field(self, field):
+        """ Remove a given field
+
+        Args:
+            field: The field to remove. ET.Element
+
+        Returns:
+            None
+        """
+        if not field or not isinstance(field, Field):
+            raise ValueError("Need to supply a field to remove element")
+
+        self._datasourceTree.getroot().remove(field.xml)
+        self._refresh_fields()
+
+    ###########
+    # Calculations
+    ###########
+    @property
+    def calculations(self):
+        """ Returns all calculated fields.
+        """
+        # TODO: There is a default [Number of Records] calculation.
+        # Should this be excluded so users can't meddle with it?
+        return {k: v for k, v in self.fields.items() if v.calculation is not None}
+
+    def add_calculation(self, caption, formula, datatype, role, type):
+        """ Adds a calculated field with the given values.
+
+        Args:
+            caption:  Caption of the new calculation. String.
+            formula:  Formula of the new calculation. String.
+            datatype:  Datatype of the new calculation. String.
+            role:  Role of the new calculation. String.
+            type:  Type of the new calculation. String.
+
+        Returns:
+            The new calculated field that was created. Field.
+        """
+        # Dynamically create the name of the field
+        name = '[Calculation_{}]'.format(str(uuid4().int)[:18])
```

Comment:
Yeah, I left it the simple way so the function is easy to grok :)

(And I copy pasted from SO)